idseq_pipeline/commands/assembly_functions.py

```diff
@@ -0,0 +1,134 @@
+        with open(pipeline_output_json) as f:
+            pipeline_output = json.load(f)
+        taxon_counts = pipeline_output['pipeline_output']['taxon_counts_attributes']
+        sorted_taxids_to_assemble = get_taxids_to_assemble(taxon_counts)
+        # Get reads for those taxids
+        output = {}
+        hit_delimiters = ['_nt', '_nr'] # annotations are like e.g. "genus_nt:543:"
+        for taxid in sorted_taxids_to_assemble:
+            pattern = '\|'.join(['%s:%s:' % (delimiter, taxid) for delimiter in hit_delimiters])
+            partial_fasta =  os.path.join(RESULT_DIR, taxid + ".fasta")
+            try:
+                subprocess.check_call("grep -A 1 --no-group-separator '%s' %s > %s" % (pattern, full_fasta, partial_fasta), shell=True)
+                output[taxid] = partial_fasta
+            except:
+                print "WARNING: taxid %s was not found in the annotated fasta" % taxid
+        # Also include the full fasta as an input to assembly
+        output['all'] = full_fasta
+        sorted_taxids_to_assemble.append('all')
+        sorted_taxids_to_assemble = [item for item in sorted_taxids_to_assemble if item in output.keys()]
+        return output, sorted_taxids_to_assemble
+
+    def length_without_newlines(sequence):
+        return len(sequence.replace("\n",""))
+
+    def read_fasta(input_fasta):
+        ''' Read fasta file into dictionary. Newlines remain part of the keys/values. '''
+        read_name = None
+        d = {}
+        read_order = []
+        with open(input_fasta, 'rb') as f:
+            for line in f:
+                if line.startswith('>'):
+                    read_name = line
+                    read_order += [read_name]
+                else:
+                    d[read_name] = d.get(read_name, '') + line
+        return d, read_order
+
+    def write_fasta(fasta_dict, read_order, output_file):
+        with open(output_file, 'wb') as f:
+            for read_name in read_order:
+                f.write(read_name + fasta_dict[read_name])
+
+    def clean_scaffolds(spades_scaffolds_fasta, min_contig_length, cleaned_fasta):
+        fasta_dict, read_order = read_fasta(spades_scaffolds_fasta)
+        for name, sequence in fasta_dict.iteritems():
+            if length_without_newlines(sequence) <= min_contig_length:
+                read_order.remove(name)
+        write_fasta(fasta_dict, read_order, cleaned_fasta)
+
+    def max_read_length(input_fasta):
+        fasta_dict, _read_order = read_fasta(input_fasta)
+        return max([length_without_newlines(sequence) for sequence in fasta_dict.values()])
+
+    def spades(input_fasta, output_fasta):
+        tmp_output_dir = input_fasta + "_temp_output"
+        tmp_output_file = tmp_output_dir + "/scaffolds.fasta"
+        try:
+            execute_command_realtime_stdout("spades.py -s %s -o %s -m 60 -t 32 --only-assembler" % (input_fasta, tmp_output_dir))
```

Comment:
nb:  I am removing `execute_command_realtime_stdout`.  Just use `execute_command`.

Reply:
Is `execute_command` still going to give realtime stdout then? For SPAdes it's quite useful as it is telling you exactly what it is doing on the reads